File: services/api/get_companies.php
<?php
/********************************************
 * Get Companies API
 * Fetches all companies from database AND file system
 ********************************************/

header("Content-Type: application/json");
header("Access-Control-Allow-Origin: *");

require_once __DIR__ . '/../shared/database/Database.php';

try {
    $db = Database::getInstance();

    // Fetch companies from database
    $dbCompanies = $db->query("
        SELECT
            c.*,
            COUNT(DISTINCT a.ad_id) as total_ads,
            COALESCE(SUM(a.views_count), 0) as total_views,
            COALESCE(SUM(a.likes_count), 0) as total_likes,
            COALESCE(SUM(a.favorites_count), 0) as total_favorites
        FROM companies c
        LEFT JOIN ads a ON c.company_slug = a.company_slug
        GROUP BY c.company_slug
        ORDER BY c.created_at DESC
    ");

    // Index database companies by slug for easy lookup
    $companiesBySlug = [];
    foreach ($dbCompanies as $company) {
        $companiesBySlug[$company['company_slug']] = $company;
    }

    // Also fetch from file system (for any companies not in database)
    $companiesDir = __DIR__ . '/../company/data/companies/';
    if (is_dir($companiesDir)) {
        $dirs = scandir($companiesDir);
        foreach ($dirs as $dir) {
            if ($dir === '.' || $dir === '..') continue;

            $companyJsonPath = $companiesDir . $dir . '/company.json';
            if (file_exists($companyJsonPath)) {
                $fileData = json_decode(file_get_contents($companyJsonPath), true);
                if ($fileData && !isset($companiesBySlug[$dir])) {
                    // Company exists in file system but not in database - add it
                    $companiesBySlug[$dir] = [
                        'company_slug' => $fileData['company_slug'] ?? $dir,
                        'company_name' => $fileData['company_name'] ?? $dir,
                        'email' => $fileData['email'] ?? '',
                        'phone' => $fileData['phone'] ?? '',
                        'status' => $fileData['status'] ?? 'active',
                        'created_at' => is_numeric($fileData['created_at']) ? $fileData['created_at'] : strtotime($fileData['created_at'] ?? 'now'),
                        'total_ads' => 0,
                        'total_views' => 0,
                        'total_likes' => 0,
                        'total_favorites' => 0
                    ];

                    // Sync to database
                    try {
                        $now = time();
                        $createdAt = is_numeric($fileData['created_at']) ? $fileData['created_at'] : strtotime($fileData['created_at'] ?? 'now');
                        $db->execute(
                            "INSERT OR IGNORE INTO companies (company_slug, company_name, email, phone, created_at, updated_at, status) VALUES (?, ?, ?, ?, ?, ?, ?)",
                            [
                                $fileData['company_slug'] ?? $dir,
                                $fileData['company_name'] ?? $dir,
                                $fileData['email'] ?? '',
                                $fileData['phone'] ?? '',
                                $createdAt,
                                $now,
                                $fileData['status'] ?? 'active'
                            ]
                        );
                    } catch (Exception $syncErr) {
                        // Ignore sync errors
                    }
                }
            }
        }
    }

    // Convert to indexed array
    $companies = array_values($companiesBySlug);

    // Calculate statistics
    $stats = [
        'total' => count($companies),
        'verified' => 0,
        'inactive' => 0,
        'suspended' => 0,
        'blocked' => 0,
        'active' => 0
    ];

    foreach ($companies as $company) {
        $status = $company['status'] ?? 'active';
        switch ($status) {
            case 'verified':
                $stats['verified']++;
                break;
            case 'inactive':
                $stats['inactive']++;
                break;
            case 'suspended':
                $stats['suspended']++;
                break;
            case 'blocked':
            case 'banned':
                $stats['blocked']++;
                break;
            case 'active':
            default:
                $stats['active']++;
                break;
        }
    }

    // Format companies data
    $formattedCompanies = array_map(function($company) {
        return [
            'company_slug' => $company['company_slug'],
            'company_name' => $company['company_name'],
            'email' => $company['email'] ?? '',
            'phone' => $company['phone'] ?? '',
            'status' => $company['status'] ?? 'active',
            'created_at' => $company['created_at'],
            'total_ads' => (int)($company['total_ads'] ?? 0),
            'total_views' => (int)($company['total_views'] ?? 0),
            'total_likes' => (int)($company['total_likes'] ?? 0),
            'total_favorites' => (int)($company['total_favorites'] ?? 0)
        ];
    }, $companies);

    // Return response
    echo json_encode([
        'success' => true,
        'companies' => $formattedCompanies,
        'stats' => $stats,
        'timestamp' => time()
    ]);

} catch (Exception $e) {
    http_response_code(500);
    echo json_encode([
        'success' => false,
        'error' => 'Failed to fetch companies',
        'message' => $e->getMessage()
    ]);
}

File: python_system/python_shared/api/get_companies.py
"""
get_companies.py - Get Companies API
Fetches all companies from database
Converted from PHP to Python
"""

from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from sqlalchemy import func, desc
import time

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent))

from models import Company, Ad
from database import get_db

router = APIRouter()


@router.get("/companies")
async def get_companies(db: Session = Depends(get_db)):
    """Fetch all companies with stats"""
    try:
        # Fetch companies with aggregated stats
        companies = db.query(
            Company,
            func.count(Ad.ad_id).label("total_ads"),
            func.coalesce(func.sum(Ad.views_count), 0).label("total_views"),
            func.coalesce(func.sum(Ad.likes_count), 0).label("total_likes"),
            func.coalesce(func.sum(Ad.favorites_count), 0).label("total_favorites")
        ).outerjoin(
            Ad, Company.company_slug == Ad.company_slug
        ).group_by(Company.company_slug).order_by(desc(Company.created_at)).all()

        # Calculate statistics
        stats = {
            "total": len(companies),
            "verified": 0,
            "inactive": 0,
            "suspended": 0,
            "blocked": 0,
            "active": 0
        }

        formatted_companies = []
        for comp, total_ads, total_views, total_likes, total_favorites in companies:
            status = comp.status or "active"

            # Update stats
            if status == "verified":
                stats["verified"] += 1
            elif status == "inactive":
                stats["inactive"] += 1
            elif status == "suspended":
                stats["suspended"] += 1
            elif status in ["blocked", "banned"]:
                stats["blocked"] += 1
            else:
                stats["active"] += 1

            formatted_companies.append({
                "company_slug": comp.company_slug,
                "company_name": comp.company_name,
                "email": comp.email or "",
                "phone": comp.phone or "",
                "status": status,
                "created_at": str(comp.created_at) if comp.created_at else "",
                "total_ads": total_ads,
                "total_views": int(total_views),
                "total_likes": int(total_likes),
                "total_favorites": int(total_favorites)
            })

        return {
            "success": True,
            "companies": formatted_companies,
            "stats": stats,
            "timestamp": int(time.time())
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "success": False,
                "error": "Failed to fetch companies",
                "message": str(e)
            }
        )


@router.get("/companies/{company_slug}")
async def get_company(company_slug: str, db: Session = Depends(get_db)):
    """Fetch single company by slug"""
    try:
        company = db.query(Company).filter(Company.company_slug == company_slug).first()

        if not company:
            return JSONResponse(
                status_code=404,
                content={"success": False, "error": "Company not found"}
            )

        # Get company stats
        stats = db.query(
            func.count(Ad.ad_id).label("total_ads"),
            func.coalesce(func.sum(Ad.views_count), 0).label("total_views"),
            func.coalesce(func.sum(Ad.likes_count), 0).label("total_likes")
        ).filter(Ad.company_slug == company_slug).first()

        return {
            "success": True,
            "company": {
                "company_slug": company.company_slug,
                "company_name": company.company_name,
                "email": company.email or "",
                "phone": company.phone or "",
                "status": company.status or "active",
                "created_at": str(company.created_at) if company.created_at else "",
                "total_ads": stats[0] if stats else 0,
                "total_views": int(stats[1]) if stats else 0,
                "total_likes": int(stats[2]) if stats else 0
            }
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )

